File: examples/governance_static_go/domain/request_changes.go
package domain

import (
	"time"
	"fmt"
)

type RequestChanges struct {
	ReviewId string `json:"review_id"`
	Notes string `json:"notes"`
}

func (c RequestChanges) CommandName() string { return "RequestChanges" }

func (c RequestChanges) Execute(repo ComplianceReviewRepository) (*ComplianceReview, *RequestedChanges, error) {
	existing, err := repo.Find(c.ReviewId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("ComplianceReview not found: %s", c.ReviewId)
	}
	existing.Notes = c.Notes
	if existing.Status != "open" {
		return nil, nil, fmt.Errorf("cannot RequestChanges: ComplianceReview is in %s state", existing.Status)
	}
	existing.Status = "changes_requested"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RequestedChanges{
		AggregateID: existing.ID,
		ReviewId: c.ReviewId,
		Notes: c.Notes,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}

File: examples/governance/compliance_domain/lib/compliance_domain/compliance_review/commands/request_changes.rb
module ComplianceDomain
  class ComplianceReview
    module Commands
      class RequestChanges
        include Hecks::Command
        emits "RequestedChanges"

        attr_reader :review_id, :notes

        def initialize(review_id: nil, notes: nil)
          @review_id = review_id
          @notes = notes
        end

        def call
          existing = repository.find(review_id)
          if existing
            unless existing.status == "open"
              raise ComplianceDomain::Error, "Cannot RequestChanges: status must be 'open', got '#{existing.status}'"
            end
            ComplianceReview.new(
              id: existing.id,
              model_id: existing.model_id,
              policy_id: existing.policy_id,
              reviewer_id: existing.reviewer_id,
              outcome: existing.outcome,
              notes: notes,
              completed_at: existing.completed_at,
              conditions: existing.conditions,
              status: "changes_requested"
            )
          else
            raise ComplianceDomain::Error, "ComplianceReview not found: #{review_id}"
          end
        end
      end
    end
  end
end
